Add convenience functions EVP_PKEY_Q_keygen(), EVP_RSA_gen(), and EVP_EC_gen()

DDvO opened the pull request:
I've just carved this out from #14664 since the key gen API can be better handled in an independent PR.

This PR solves some part of #14628. It does the following.
* add simple-to-use asymmetric key generation functions:
  ~the general function `EVP_keygen()`,~ the shorthand `EVP_PKEY_Q_keygen`, 
  and two macros for the most common special cases: `EVP_RSA_gen()` and `EVP_EC_gen()`.
* export aux function~s~ `OSSL_EC_curve_nid2name()` ~and OSSL_EC_curve_name2nid()~
* improve doc and deprecation info on RSA and EC key generation/management functions
* make `OSSL_PKEY_PARAM_RSA_BITS` consistently `size_t`
* Constify `EVP_PKEY_CTX_set_params()`, `EVP_PKEY_CTX_{set,get}table_params()`, etc.

- [x] documentation is added or updated
- [x] tests are added or updated

mattcaswell commented:
I've not properly reviewed this at this stage. But it seems wrong to me that these functions are in the OSSL_ namespace. They should surely be EVP_. Also why do we need the `OSSL_EC_curve_nid2name`/`OSSL_EC_curve_name2nid()` functions at all?

DDvO commented:
> I've not properly reviewed this at this stage. But it seems wrong to me that these functions are in the OSSL_ namespace. They should surely be EVP_. 

No problem, will change them to `EVP_keygen()`, `EVP_RSA_keygen()`, and `EVP_EC_keygen()`.
(I had thought that  `EVP_keygen()` was already taken, but I had confused that with  `EVP_PKEY_keygen()`.)

> Also why do we need the `OSSL_EC_curve_nid2name`/`OSSL_EC_curve_name2nid()` functions at all?

In my original PR that included these functions, the EC keygen function had an NID parameter for the curve, so to support that I had to export `OSSL_EC_curve_name2nid()`.
Now, conversely, the EC genkey function has a string param for giving the curve, so I thought it may help to add some flexibility such that users can specify curves also using their NID.
Or is this not of interest?

t8m commented:
We should not add functionality that is supposed to go away. And NIDs are definitely supposed to go away given they are something that has to be hardcoded in the libcrypto.

t8m commented:
All the functions should have OSSL_PARAMS and propq argument. This at least makes them somewhat universally usable.

DDvO commented:
> We should not add functionality that is supposed to go away. And NIDs are definitely supposed to go away given they are something that has to be hardcoded in the libcrypto.

So I'll make the two helper functions internal again.

> All the functions should have OSSL_PARAMS and propq argument. This at least makes them somewhat universally usable.

The fully general
```
 EVP_PKEY *EVP_keygen(OSSL_LIB_CTX *libctx, const char *name,
                      const char *propq, OSSL_PARAM *params);
```
already has them, and the `EVP_RSA_keygen()` and `EVP_RSA_keygen()` are really meant as very simple functions that cover the gross of RSA and EC keygen.
Sure I can add the two more parameters, yet for presumably 98% of their use the `propq` and `params` arguments will be NULL.
Moreover, adding the `params` parameter would complicate their implementation because unfortunately in C  one cannot simply concatenate those params with the one RSA / EC parameter these functions provide internally.
Still wanna have that?

t8m commented:
OK, I give up on the params argument then, but please add the `propq` at least.

slontis commented:
I can see the mac and digest one-shot functions being helpful for some users...
keygen is 5 lines of code - is this maybe taking the helpers a bit too far?

DDvO commented:
> I can see the mac and digest one-shot functions being helpful for some users...
> keygen is 5 lines of code - is this maybe taking the helpers a bit too far?

If an application just needs a key for whatever, why bother its developer with having to find a code snippet/example how to do it, with copying it in, polluting the actual application code, and maintaining it?
Just have a look how the new convenience functions simplify even our test code.

DDvO commented:
BTW, if there were much such convenience functions also for other topics,
quite some share of the OpenSSL code would become more concise and much easier to read and maintain.

levitte commented on include/openssl/ec.h:
```diff
@@ -84,6 +84,9 @@ typedef enum {
 } point_conversion_form_t;
 
+const char *OSSL_EC_curve_nid2name(int nid);
+int OSSL_EC_curve_name2nid(const char *name);
```
I can understand the purpose of `OSSL_EC_curve_nid2name()`, for apps that use NIDs and want the corresponding name to use with `OSSL_EC_keygen()`...  but the other one, no.

levitte commented on doc/man3/RSA_generate_key.pod:
```diff
@@ -2,32 +2,37 @@
 
  #include <openssl/rsa.h>
 
+ EVP_PKEY *OSSL_RSA_keygen(OSSL_LIB_CTX *libctx, int bits);
```
Would it be interesting to have it also take a `primes`, or to have a second function for multiprime?  So either change to this:


``` C
EVP_PKEY *OSSL_RSA_keygen(OSSL_LIB_CTX *libctx, int bits, int primes);
```
 or add this:

``` C
EVP_PKEY *OSSL_RSA_multiprime_keygen(OSSL_LIB_CTX *libctx, int bits, int primes);
```

slontis replied:
Yep and this is entirely why I am not sure this is a good idea..
It is a potential API explosion.
I have no problem with the generic one shot OSSL_keygen() though (except for its name) - maybe EVP_PKEY_keygen_something??

slontis replied:
All this tells me is that people are scared of OSSL_PARAMS - which was supposed to be a replacement for this kind of API bloat (at least I thought it was).

richsalz replied:
> All this tells me is that people are scared of OSSL_PARAMS 

You can't really say there's an opinion about `OSSL_PARAMS` based on this one PR. :)

slontis replied:
I may form this opinion based on whether we need this API that just sets a single parameter and then calls a one shot function.
Even 'bits' is an optional parameter for keygen since it uses a default if it is not specified..

DDvO replied:
>What about DSA, DH, ED25519, ED448, X25519, X448, SM2?

ED25519, ED448, X25519, X448 are covered easily by
```
    OSSL_keygen(NULL, "ED25519", NULL, NULL);
```
and the like. I've meanwhile made this clear in `EVP_PKEY-X25519.pod`.

I see not much point for special convenience functions for DSA and DH
since there are a lot of variations and these key types have become unpopular.
Neither for for SM2 and the like since they are rarely used.

DDvO replied:
> Would it be interesting to have it also take a primes, or to have a second function for multiprime?

I do not think that it is worth adding more coverage for special forms of RSA key generation etc. because they are pretty rarely used.

DDvO replied:
> It is a potential API explosion.

As mentioned, I believe it is worth just adding rather simple `EVP_RSA_keygen()` and `EVP_EC_keygen()` convenience functions because they already cover the commonly used cases.

> I have no problem with the generic one shot OSSL_keygen() though (except for its name) - maybe EVP_PKEY_keygen_something??

It's meanwhile `EVP_keygen()`. 
As mentioned, `EVP_PKEY_keygen()` is already taken.

DDvO replied:
> All this tells me is that people are scared of OSSL_PARAMS

I would not be afraid of using them, 
but for common cases I consider them plainly inconvenient to use.

levitte commented on crypto/ec/ec_lib.c:
```diff
@@ -1734,3 +1734,13 @@ EC_GROUP *EC_GROUP_new_from_params(const OSSL_PARAM params[],
+    OSSL_PARAM params[] =
+        { OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME,
+                                           (char *)curve, 0),
+          OSSL_PARAM_END };
```
I was surprised to see this come through the CIs...  but it seems that we have lost the build that checked if the source can be built with `-ansi`, that explains it.  Unfortunately, this construct won't compile...  (I've been hit more than once)

DDvO replied:
This construct is used already several time in crypto and test, apparently without problems.

levitte reviewed (changes requested):
What about DSA, DH, ED25519, ED448, X25519, X448, SM2?

slontis commented on crypto/evp/evp_lib.c:
```diff
@@ -1024,3 +1026,17 @@ int EVP_PKEY_CTX_get_group_name(EVP_PKEY_CTX *ctx, char *name, size_t namelen)
+    EVP_PKEY *pkey = NULL;
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_from_name(libctx, name, propq);
+
+    if (ctx != NULL && EVP_PKEY_keygen_init(ctx) > 0
```
Move && to next line?

DDvO replied:
Done

slontis commented on doc/man3/RSA_generate_key.pod:
```diff
@@ -2,32 +2,37 @@
 
 =head1 NAME
 
+OSSL_RSA_keygen,
```
I am not sure this belongs in this file..
It should probably be in L<EVP_PKEY-RSA(7)> and just be pointing to there from this file..

DDvO replied:
`L<EVP_PKEY-RSA(7)>` does not document actual function names,
and people would expect RSA key gen functions described in `RSA_generate_key.pod`.
Moreover, it make sense to describe the new function there as a simple replacement for the most common use of the deprecated functions. 

BTW, the function meanwhile is called `EVP_RSA_keygen`.

Good point though to add an example using the new simple function to `L<EVP_PKEY-RSA(7)>`:

    EVP_RSA_keygen(NULL, 4096, NULL);

DDvO replied:
Likewise added
```
    EVP_EC_keygen(NULL, "P-256", NULL);
```
to `EVP_PKEY-EC.pod`.

slontis replied:
> expect RSA key gen functions described in `RSA_generate_key.pod`.
> Moreover, it make sense to describe the new function there as a simple replacement for the most common use of the deprecated functions.

Maybe EVP_PKEY_gen.pod is a better place for the functions with a linl in RSA_generate_key.pod?

DDvO replied:
Would be an option.
Currently  EVP_PKEY_gen.pod contains the doc of `EVP_keygen()`
and contains
```
 =head1 SEE ALSO

L<EVP_RSA_keygen(3)>, L<EVP_EC_keygen(3)>,
```
which I find slightly more adequate.

slontis replied:
This would be less of a problem with non pod-like documentation that had more logical structure :)

levitte replied:
@slontis, Unix man page section 3 is for library functions.  Documenting the functions from this PR in section 7 is simply wrong in that sectioning.

That being said, if one can't find a suitable file, *create a new file*!

DDvO commented:
@t8m,

> No problem, will change them to `EVP_keygen()`, `EVP_RSA_keygen()`, and `EVP_EC_keygen()`.

Done.

> but please add the `propq` at least.

Done.

>> Now, conversely, the EC genkey function has a string param for giving the curve, so I thought it may help to add some flexibility such that users can specify curves also using their NID.

Since @levitte also believes it can be helpful, I kept `OSSL_EC_curve_nid2name()` public.

slontis commented:
You might want to reword the commit at some stage when the names are agreed on :)

DDvO commented:
> You might want to reword the commit at some stage when the names are agreed on :)

Argh, I forgot to update the commit message as well. Will be part of the next push.

DDvO commented:
As decided today, I've added `EVP_Q_keygen()` and used it for defining macros
`EVP_rsa_keygen()` and `EVP_ec_keygen()`. 
BTW, should those really take the `libctx` and `propq` parameters, as the most likely will be NULL?

Also have taken the liberty to remove the 'hold: need otc decision' label myself, and rebased this PR.

t8m commented:
> BTW, should those really take the `libctx` and `propq` parameters, as the most likely will be NULL?

The EVP_Q_keygen() should be better named EVP_PKEY_Q_gen() or EVP_Q_pkey_gen(). It should definitely have those parameters as especially all the general purpose libraries that use OpenSSL should sooner or later start using non-NULL libctx. Of course applications can keep using the default context but...

For the EVP_rsa_keygen (maybe EVP_Q_rsa_keygen() ?) I do not know. IMO it wouldn't hurt to have these parameters but I probably would not insist on that.

paulidale commented on crypto/evp/evp_lib.c:
```diff
@@ -1046,3 +1048,35 @@ int EVP_PKEY_CTX_get_group_name(EVP_PKEY_CTX *ctx, char *name, size_t namelen)
+    return pkey;
+}
+
+EVP_PKEY *EVP_Q_keygen(OSSL_LIB_CTX *libctx, const char *type,
```
I'm wondering if one of these two APIs abse on this function prototype might be more friendly and future proof:

```c
EVP_PKEY *EVP_Q_keygen(OSSL_LIB_CTX *libctx,
                       const char *type,
                       const char *propq,
                       ...)
```

The first idea being that the `type` determines the varargs component.  Thus, e.g., RSA would have an `unsigned int` argument specifying the key size and the wrapper could be:
```c
#define EVP_rsa_keygen(libctx, propq, bits) \
    EVP_Q_keygen(libctx, "RSA", propq, (unsigned int)(0 + bits))
```

EC would take a string vararg which could be macroed:
```c
#define EVP_ec_keygen(libctx, propq, curve) \
    EVP_Q_keygen(libctx, "EC", propq, curve)
```

We'd retain some semblance of type checking via the macros and wouldn't be limited when an algorithm needs two arguments or a floating point number.


My other thought is to have name/value pairs in the varargs part.  These continue until a NULL name is encountered which can be checked by some compilers via the `sentinel` attribute.

RSA would thus be:
```c
#define EVP_rsa_keygen(libctx, propq, bits) \
    EVP_Q_keygen(libctx, "RSA", propq, "size", (unsigned int)(0 + bits), NULL)
```

and EC:
```c
#define EVP_ec_keygen(libctx, propq, curve) \
    EVP_Q_keygen(libctx, "EC", propq, "curve", curve, NULL)
```

Once we know the type, the settable params become available and they can be used to determine the type of the argument following the name (there would need to be some restrictions).  We could also, e.g. allow "params" to be followed by an OSSL_PARAM array argument.

DDvO replied:
Interesting thoughts.
Although this PR already introduces the fully general one-shot keygen function
```
EVP_PKEY *EVP_keygen(OSSL_LIB_CTX *libctx, const char *name,
                     const char *propq, OSSL_PARAM *params);
```
I like the idea of making `EVP_Q_keygen()` (or `EVP_PKEY_Q_gen()`, as it is meanwhile called) more elegant and somewhat more flexible. So changed it to
```
EVP_PKEY *EVP_PKEY_Q_gen(OSSL_LIB_CTX *libctx, const char *propq,
                          const char *type, ...);
```
and defined the two convenience macros as
```
# define EVP_RSA_gen(bits) \
    EVP_PKEY_Q_gen(NULL, NULL, "RSA", (unsigned int)(0 + (bits)))
# define EVP_EC_gen(curve) \
    EVP_PKEY_Q_gen(NULL, NULL, "EC", (char *)(strstr(curve, "")))

```
where I added `strstr(curve, "")` in order to provide type checking also for the string variadic arg.

t8m replied:
Hmm `EVP_keygen` should be probably named `EVP_PKEY_gen_with_params` or maybe even better `EVP_PKEY_gen_from_params`.

paulidale replied:
Using `strstr(curve, "")` is clever.  Another possibly would be `strchr(curve, *curve)`.  I don't see much difference between the two, so this isn't a request to change.  My initial thought for checking the string type was `curve ""` which doesn't work for pointers so I skipped it.  I was in a rush so didn't dwell over other possibilities.

I agree with @t8m that the general function should start with `EVP_PKEY_`.

DDvO replied:
> Using `strstr(curve, "")` is clever. Another possibly would be `strchr(curve, *curve)`. I don't see much difference between the two, so this isn't a request to change. My initial thought for checking the string type was `curve ""` which doesn't work for pointers so I skipped it. I was in a rush so didn't dwell over other possibilities.

I checked also other possibilities and found that  `strstr(curve, "")` is best because, as opposed to alternatives, it is linear in the `curve` parameter, so macro expansion does not cause problems in case the actual argument filled in for `curve` has side effects.

DDvO replied:
> Hmm EVP_keygen should be probably named EVP_PKEY_gen_with_params or maybe even better EVP_PKEY_gen_from_params.

> I agree with @t8m that the general function should start with `EVP_PKEY_`.

I've changed the name of `EVP_keygen()` to `EVP_PKEY_gen()` (rather than a more complicated name) for brevity and consistency with `EVP_PKEY_Q_gen()`.
To this end I renamed `EVP_PKEY_gen()`, which had been introduced for 3.0 earlier, to  `EVP_PKEY_generate()` (this more low-level function is anyway not used often).

t8m replied:
EVP_PKEY_gen() is truly unfortunate name for this quick generation function and I realize that names proposed by myself above were not good either. The reason is it is not clear from the name whether a key is generated or domain parameters are generated. With the EVP_PKEY_Q_gen (which should be actually EVP_PKEY_Q_keygen unless you plan to somehow support generating domain parameters) - is there actually any reason to have a public one-shot keygen function that uses the OSSL_PARAMS? I'd just drop it. There is IMO too much choice. You have the EVP_PKEY_Q_keygen for the simple stuff and the existing functions for the rest.

And I am fine with renaming EVP_PKEY_gen() to  EVP_PKEY_generate(). IMO the EVP_PKEY_gen() name is ugly.

DDvO replied:
Ok, made `EVP_PKEY_gen()` static (as `evp_pkey_keygen()`)
and renamed `EVP_PKEY_Q_gen()` to `EVP_PKEY_Q_keygen()`.

DDvO commented:
> The EVP_Q_keygen() should be better named EVP_PKEY_Q_gen() or EVP_Q_pkey_gen().

Changed to `EVP_PKEY_Q_gen()`.

> It should definitely have those parameters as especially all the general purpose libraries that use OpenSSL should sooner or later start using non-NULL libctx. 

Yes; my suggestion to drop them was just meant for the simple RSA and EC key gen macros.

> Of course applications can keep using the default context but...
> For the EVP_rsa_keygen (maybe EVP_Q_rsa_keygen() ?) I do not know.
> IMO it wouldn't hurt to have these parameters but I probably would not insist on that.

I've just dropped them because if needed, `EVP_PKEY_Q_gen()` can be used rather conveniently.
And I've tentatively shortended the names of the convenience macros to `EVP_RSA_gen()` and `EVP_EC_gen()`.

petrovr commented:
Nothing rationale with function per "key type". With 1.0 OpenSSL goes into key "generalisation". This is huge step backward. 
So what to expect "SM2 key gen" "SM3 key gen" "SM13 key gen" "SM53 key gen" "SM97 key gen" "SM2021  gen" ?

DDvO commented:
@petrovr, I do not understand what you mean - please expand.

Note that this PR takes nothing away from the new very general and future-proof API.
It just adds some pragmatic convenience functions for the most commonly used cases of asymmetric key generation: RSA keys of given size and EC keys on a given named curve.

DDvO commented:
Rebased to fix merge conflicts.
Can this be approved now?

DDvO commented:
I wonder why the 3.0.0-beta1 milestone was gone - this needs to get in before API/feature freeze.
So re-added it.

slontis commented:
It means OTC is not doing this task.. If someone else does it though it may go in..

slontis commented on CHANGES.md:
```diff
@@ -771,6 +778,9 @@ OpenSSL 3.0
    L<EVP_PKEY_decrypt(3)>.
+   For replacement of the functions manipulating the RSA objects
+   see the L<EVP_PKEY-RSA(7)> manual page.
+   A simple way of generating RSA keys is L<EVP__RSA_gen(3)>.
```
Fix Double underscore

slontis commented on crypto/evp/evp_lib.c:
```diff
@@ -1104,3 +1106,56 @@ int EVP_PKEY_CTX_get_group_name(EVP_PKEY_CTX *ctx, char *name, size_t namelen)
+ * of algorithm type I<name> (e.g., C<RSA> or C<EC>).
+ * The library context I<libctx> and property query I<propq>
+ * are used when fetching algorithms from providers.
+ * The I<params> specify algorithm-specific paramaters
```
parameters?

slontis commented on crypto/evp/evp_lib.c:
```diff
@@ -1104,3 +1106,56 @@ int EVP_PKEY_CTX_get_group_name(EVP_PKEY_CTX *ctx, char *name, size_t namelen)
+
+    va_start(args, type);
+
+    if (strcmp(type, "RSA") == 0) {
```
do we care about case here?

t8m replied:
We shouldn't.

slontis commented on doc/man3/RSA_generate_key.pod:
```diff
@@ -2,36 +2,40 @@
 RSA_generate_key_ex() generates a 2-prime RSA key pair and stores it in the
 B<RSA> structure provided in B<rsa>. The pseudo-random number generator must
-be seeded prior to calling RSA_generate_key_ex().
+be seeded prior< to calling RSA_generate_key_ex().
```
extra <

DDvO commented:
Thanks @slontis for your review pass. I've just fixed all nits reported.

Anything more to do, or how about approving this?

slontis commented on crypto/evp/evp_lib.c:
```diff
@@ -14,6 +14,15 @@
 #include "internal/deprecated.h"
 
 #include <stdio.h>
+#include <string.h>
```
Please just do this..

#include "e_os.h" /* strcasecmp */

Propagating this block of logic doesnt seem like a good idea.

DDvO replied:
Yeah, I had copied this over from `conf_def.c` but already suspected that there should be a better way.
Changes as suggested in both files.

slontis replied:
At some point we may fix this mess related to strcasecmp..

DDvO commented:
Rebased again to fix merge conflicts.
Ping @openssl/otc for approval.

DDvO commented:
check_update CI failure appears unrelated.

slontis commented on doc/man3/EVP_PKEY_keygen.pod:
```diff
@@ -87,13 +91,25 @@ and retrieve an opaque pointer. This can be used to set some application
+can be provided fetching algorithms from providers.
+If I<type> is C<RSA>, the following unsigned integer parameter is used to
+specify the size of RSA keys.
+If I<types> is C>EC>, the I<name> is used to specify the name of an EC curve.
```
<>
I thought we used B<> in other places?

DDvO replied:
Function parameters were formerly marked by `B< >` but nowadays we use `I< >`,
while for type names like `EVP_PKEY_CTX` we use `B< >`.

paulidale replied:
The `C>EC>` looks wrong.

DDvO replied:
Yes, I also noticed this, fixed.

slontis commented on doc/man3/RSA_generate_key.pod:
```diff
@@ -2,36 +2,40 @@
-Applications should instead use L<EVP_PKEY_keygen_init(3)> and
-L<EVP_PKEY_keygen(3)>.
+EVP_RSA_gen() generates a new RSA key pair with modulus size B<bits>.
+The pseudo-random number generator must be seeded before using this macro.
```
@paulidale does this seeded statement still apply?

paulidale replied:
No.  We self seed as required.

DDvO replied:
So removed this remark about seeding, also in other such places.

slontis commented on doc/man7/EVP_PKEY-EC.pod:
```diff
@@ -159,6 +159,10 @@ An B<EVP_PKEY> context can be obtained by calling:
 An B<EVP_PKEY> ECDSA or ECDH key can be generated with a "P-256" named group by
 calling:
 
+    EVP_EC_gen("P-256");
```
pkey =

DDvO replied:
Added

slontis commented on doc/man7/EVP_PKEY-RSA.pod:
```diff
@@ -202,14 +202,18 @@ An B<EVP_PKEY> context can be obtained by calling:
-An B<RSA> key can be generated like this:
+An B<RSA> key can be generated simply like this:
+
+    EVP_RSA_gen(4096);
```
pkey =

DDvO replied:
Added

slontis commented on doc/man7/EVP_PKEY-X25519.pod:
```diff
@@ -86,23 +86,11 @@ An B<EVP_PKEY> context can be obtained by calling:
-    EVP_PKEY_keygen_init(pctx);
-    EVP_PKEY_gen(pctx, &pkey);
-    EVP_PKEY_CTX_free(pctx);
+    OSSL_keygen(NULL, "ED25519", NULL, NULL);
```
OSSL_ ??

DDvO replied:
Oops, forgot to update this. Done.

slontis commented on doc/man7/EVP_PKEY-X25519.pod:
```diff
@@ -86,23 +86,11 @@ An B<EVP_PKEY> context can be obtained by calling:
-    EVP_PKEY_keygen_init(pctx);
-    EVP_PKEY_gen(pctx, &pkey);
-    EVP_PKEY_CTX_free(pctx);
+    OSSL_keygen(NULL, "ED25519", NULL, NULL);
```
pkey =

DDvO replied:
fixed

slontis commented on doc/man7/EVP_PKEY-X25519.pod:
```diff
@@ -86,23 +86,11 @@ An B<EVP_PKEY> context can be obtained by calling:
-    EVP_PKEY_keygen_init(pctx);
-    EVP_PKEY_gen(pctx, &pkey);
-    EVP_PKEY_CTX_free(pctx);
+    OSSL_keygen(NULL, "X25519", NULL, NULL);
```
same here

DDvO replied:
fixed

slontis commented on doc/man7/EVP_PKEY-RSA.pod:
```diff
@@ -110,7 +110,7 @@ When generating RSA keys, the following key generation parameters may be used.
 =over 4
 
-=item "bits" (B<OSSL_PKEY_PARAM_RSA_BITS>) <unsigned integer>
+=item "bits" (B<OSSL_PKEY_PARAM_RSA_BITS>) <size_t>
```
Why?

This maps to OSSL_PKEY_PARAM_BITS which is an integer..
OSSL_PKEY_PARAM_MAX_SIZE & OSSL_PKEY_PARAM_SECURITY_BITS are also integers, so it seems strange to change this one..

t8m replied:
Also the integer parameters are not really sized. It is `unsigned integer` and not `unsigned int` for a reason.

paulidale replied:
`unsigned integer` is correct here.

DDvO replied:
There were inconsistencies w.r.t. the type of `OSSL_PKEY_PARAM_RSA_BITS`, which I resolved in favor of `size_t`.
See e.g., `providers/implementations/keymgmt/rsa_kmgmt.c`

paulidale replied:
I'm not too surprised there are inconsistencies :(

t8m replied:
The integer parameters themselves do not have a size. The get/set functions automatically handle size conversions.

paulidale replied:
The documentation doesn't specify C types, it specifies OSSL_PARAM types.  The type the C code uses is separate.

There are two `size_t` mistakes in the documentation which I'll fix.

t8m replied:
@paulidale  Already submitted #15178

t8m replied:
> There are two `size_t` mistakes in the documentation which I'll fix.

I only see one size_t in parameter list though.

DDvO replied:
So reverted to `unsigned integer`.

slontis commented on doc/man3/RSA_generate_key.pod:
```diff
@@ -2,41 +2,42 @@
 RSA_generate_key_ex() generates a 2-prime RSA key pair and stores it in the
-B<RSA> structure provided in B<rsa>. The pseudo-random number generator must
-be seeded prior to calling RSA_generate_key_ex().
+B<RSA> structure provided in B<rsa>.
```
I<rsa>

DDvO replied:
Fixed, as well as many other such cases in the file.

slontis commented on doc/man3/RSA_generate_key.pod:
```diff
@@ -2,41 +2,42 @@
+B<RSA> structure provided in B<rsa>.
 
 RSA_generate_multi_prime_key() generates a multi-prime RSA key pair and stores
 it in the B<RSA> structure provided in B<rsa>. The number of primes is given by
```
B<rsa> and B<primes> should be I<> ?

DDvO replied:
Yes, only for type and constant names (except NULL) the `B< >` remains.

slontis commented on providers/implementations/keymgmt/rsa_kmgmt.c:
```diff
@@ -309,7 +309,7 @@ static int rsa_get_params(void *key, OSSL_PARAM params[])
 
     if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_BITS)) != NULL
-        && (empty || !OSSL_PARAM_set_int(p, RSA_bits(rsa))))
+        && (empty || !OSSL_PARAM_set_size_t(p, RSA_bits(rsa))))
```
This should not be changed here..

DDvO replied:
Reverted.

DDvO commented:
Thanks @slontis for your further comments; done all fixes.

DDvO commented:
Rebased again to update FIPS checksums.
I wonder why this is needed each time - wouldn't it be sufficient to do that for releases?

mattcaswell commented:
> @openssl/otc ? - This doesnt seem to be a very workable solution currently (i.e: This is way more intrusive to workflow than libcrypto.num). Per release would make sense. We can see what files changed per release - and then do a file history to see what changed?

I agree, although it would be useful to have it flagged up in some way that a PR is making changes inside the boundary. That will become important once we are in a released state.

t8m commented:
> I agree, although it would be useful to have it flagged up in some way that a PR is making changes inside the boundary. That will become important once we are in a released state.

Have a job similar to the CLA check job that would set a label on the PR?

The check_update job could first run on no-fips build which would avoid changing the checksums. And then it could run with enable-fips but it would not fail if checksum is changed and just flag the failure?

t8m commented:
Or we could split the checksum job out of make update completely.

mattcaswell commented:
> But to correctly flag the failure we would still have to keep the checksums up-to-date when merging PRs!

Unless the job that does the checking updates the checksums before applying the PR, then applies the PR, then checks the checksums again.

DDvO commented:
Rebased to fix various small merge conflicts (not only FIPS checkums in this case).

slontis reviewed (approved):
Thanks for persisting with this..

DDvO commented:
Rebased to fix FIPS checksum conflict and merged - thanks @slontis et al.

mattcaswell commented on doc/man7/EVP_PKEY-X25519.pod:
```diff
@@ -84,25 +84,11 @@ An B<EVP_PKEY> context can be obtained by calling:
-    EVP_PKEY *pkey = NULL;
-    EVP_PKEY_CTX *pctx =
-        EVP_PKEY_CTX_new_from_name(NULL, "X25519", NULL);
+    pkey = EVP_Q_keygen(NULL, NULL, "X25519");
```
Sorry the post-merge comment - but shouldn't this have been `EVP_PKEY_Q_keygen`?

t8m replied:
Yes, I've asked for that as well.

DDvO replied:
Good spotting!
Fixup is in #15226.
